File: facultate/an II/af/examen/parcurgeri/Parcurgeri/mod_parcurgeri.py
import mod_liste_adiacenta as mla
from collections import deque

def BF(n,la,s=1):
        s-=1  
        viz=[0]*n
        tata=[None]*n
        d=[None]*n
        q = deque()
        parc_bf=[]
        q.append(s)
        viz[s]=1
        d[s]=0
        while len(q)>0:
            x=q.popleft()
            parc_bf.append(x+1)
            for y in la[x]:
                if viz[y]==0:
                      viz[y]=1
                      tata[y]=x
                      d[y]=d[x]+1
                      q.append(y)
                
        return parc_bf,d,tata 
 
def DF(n,la,s=1):
    s-=1;  
    viz=[0]*n
    tata=[None]*n
    d=[None]*n
    parc_df=[]
    d[s]=0
    DFs(n,la,s,viz,tata,d,parc_df)
    return parc_df,d,tata

def DFs(n,la,x,viz,tata,d,parc_df): 
    viz[x]=1
    parc_df.append(x+1)
    for y in la[x]:
        if viz[y]==0:
             tata[y]=x
             d[y]=d[x]+1
             DFs(n,la,y,viz,tata,d,parc_df)
             
def test():
    #test modul
    print("orientat:")
    n,la=mla.citire(True,"grafor.in")
    mla.afisare_liste(la)
    parc_bf,d,tata = BF(n,la,1)
    print("BF(1):",parc_bf,d,tata)
    parc_df,d,tata = DF(n,la,1)
    print("DF(1):",parc_df,d,tata)

    print("neorientat:")
    n,la=mla.citire(False,"graf.in")
    mla.afisare_liste(la)
    parc_bf,d,tata = BF(n,la,1)
    print("BF(1):",parc_bf,d,tata)
    parc_df,d,tata = DF(n,la,1)
    print("DF(1):",parc_df,d,tata)

File: facultate/an II/af/examen/parcurgeri/Parcurgeri/mod_parcurgeri.cpp
#include "mod_liste_adiacenta.cpp"
#include<fstream>
#include<vector>
#include<iostream>
#include<queue>
using namespace std;


void DFs(int n, vector<int> *la,int x, int *viz, vector<int> &parc_df,int *&d,int *&tata){
    viz[x]=1;
    parc_df.push_back(x+1);
    for (int i=0;i<la[x].size();i++){
		int y=la[x][i];
        if (viz[y]==0){
		     tata[y]=x;
             d[y]=d[x]+1;
             DFs(n,la,y,viz,parc_df,d,tata);
        }
    }
}

void DF(int n, vector<int> *la, vector<int> &parc_df,int *&d,int *&tata, int s=1){
    int *viz=new int[n];
    tata=new int[n+1];
	d=new int[n+1];
	parc_df.clear();
	for(int i=0;i<n;i++){
		tata[i]=d[i]=-1;
		viz[i]=0;
	}
	s--;
    d[s]=0;
    DFs(n,la,s,viz,parc_df,d,tata);
}
             
void BF(int n, vector<int> *la, vector<int> &parc_bf,int *&d,int *&tata, int s=1)
{
	parc_bf.clear();
	int i,*viz=new int[n];
	tata=new int[n+1];
	d=new int[n+1];
	for(i=0;i<n;i++){
		tata[i]=d[i]=-1;
		viz[i]=0;
	}
	s--;
	queue<int> c;
	c.push(s);
	viz[s]=1;
	d[s]=0;
	
	while(c.size()>0){
		int x=c.front();//x- nodul curent
		c.pop();
		parc_bf.push_back(x+1);
		for (i=0;i<la[x].size();i++){
			int y=la[x][i];//pentru toti vecinii y ai nodului curent x
			if(viz[y]==0)
				{ 
					viz[y]=1;
					tata[y]=x; 
					d[y]=d[x]+1;
					c.push(y);
				}
		}
	}
}

void afis(int n,vector<int> parc_bf,int *d,int *tata){
	int i;
	for(i=0;i<parc_bf.size();i++)
	    cout<<parc_bf[i]<<" ";
	cout<<endl;
	for(i=0;i<n;i++)
	    cout<<d[i]<<" ";
	cout<<endl;  
	for(i=0;i<n;i++)
	    cout<<tata[i]<<" ";
	cout<<endl;    
}
int main(){
	vector<int> *la,parc_bf,parc_df;
	int n,m,s;
	int i,j;
	int  *tata,*d;
    
    cout<<"neorientat"<<endl;
	citire(n,la );
    afisare_liste(n,la);
    s=1;
   	BF(n,la,parc_bf,d,tata,s);
   	afis(n,parc_bf,d,tata);
   	DF(n,la,parc_df,d,tata,s);
   	afis(n,parc_df,d,tata);
   	
   	cout<<"orientat"<<endl;
	citire(n,la,1,"grafor.in" );
    afisare_liste(n,la);
    s=1;
   	BF(n,la,parc_bf,d,tata,s);
	afis(n,parc_bf,d,tata);
	DF(n,la,parc_df,d,tata,s);
   	afis(n,parc_df,d,tata);
}
